game/scripts/vscripts/items/sonic.lua

```diff
@@ -0,0 +1,64 @@
+item_sonic = class(ItemBaseClass)
+
+LinkLuaModifier("modifier_item_sonic", "items/sonic.lua", LUA_MODIFIER_MOTION_NONE)
+LinkLuaModifier("modifier_sonic_fly", "modifiers/modifier_sonic_fly.lua", LUA_MODIFIER_MOTION_NONE)
+
+function item_sonic:GetIntrinsicModifierName()
+  return "modifier_sonic_fly"
+end
+
+function item_sonic:OnToggle()
+  local caster = self.GetCaster()
+  local state = self.GetToggleState()
+  local ability = self.GetAbility()
+
+  if state == true then
+    caster.AddNewModifier(caster, self, "modifier_sonic_fly", {})
+  else
+    caster.RemoveModifierByName("modifier_sonic_fly")
+  end
+  ability:StartCooldown(ability:GetCooldown(ability:GetLevel()))
+end
+
+function item_sonic:OnOwnerDied()
```

Comment:
Modifiers are automatically removed when you die unless the modifiers says that it should stay active or is an intrinsic modifier. You probably needed this while testing because you accidentally set the fly to be intrinsic.

Reply:
Intrinsic modifiers are still removed on death if they're not set to not RemoveOnDeath, but they get reapplied on respawn. Consequently, pretty much all intrinsic modifiers ought to be unpurgable and RemoveOnDeath false. Though the latter is typically only important in somewhat corner cases.